### ISSUE TITLE
Callback when connection is ready (was: Performance regression?)

### ISSUE THREAD

**Issue by bkchr** [private-octopus/picoquic#489]
I wanted to update my bindings from https://github.com/bkchr/picoquic/commit/afe4397663e6c865718638a0fe139c71e2296729
to latest master. But with latest master my tests are running really slow. I did not had any time yet to look deeper into it, do you maybe implicitly changed something that I miss?

**Comment by huitema** [comment]
@bkchr I don't see performance regression in the current set of tests -- the execution times remained pretty much the same. There was a fix to the newreno implementation that made those tests faster, and after that perf remained pretty much constant.

**Comment by bkchr** [comment]
Okay, sorry for the noise. This was a problem with my rust implementation. I wait until a connection is in ready state, before giving it to the user. In older picoquic versions that does indeed work, because I assume that wake up times was calculate different. Nevertheless, I now switched to `client_ready_start` as the state to wait for :)

**Comment by huitema** [comment]
@bkchr that was the reason why I asked you to review the PR that introduced these two new states, on server and client state. The new state more or less state that this end is ready, but there is no certainty that the other end also is. Specifically:

1) picoquic_state_server_false_start: the server has completed the handshake and computed the session keys, but has not received confirmation that the client is ready. There might be a man-in-the-middle. The server will only move to "picoquic_state_ready" when it receives the OK from the TLS stack, i.e. after the TLS stack has successfully processed the "Client Finished" message.

2) picoquic_state_client_ready_start: the client has completed the handshake and computed the session keys, but does not know yet whether the server is ready. The connection is secure, because the client's TLS stack has processed the "server finished" message, but the client may still need to repeat the handshake messages until the server receives the "Client Finished" message. The client moves to "picoquic_state_ready" when it receives the first acknowledgement on a 1RTT protected packet, which proves that the server can properly receive and process encrypted message.

Once in the "picoquic_state_ready" stage, the peers know that they don't need to process any further Handshake message, and can use a more streamlined processing. The stack waits to reach that stage to enable some functions like key rotation.

**Comment by bkchr** [comment]
I also looked into the code and the `picoquic_ready_start` is only set in the sender, after checking that the corresponding data was received. I wondered why that is not done in the receive frame function. I tested it and put the corresponding code that switch into the ready state into the receive frame function. This worked as well with my implementation of checking for the ready state.
If you also think that the state switch could be put into the receive frame function, I could open a pr.
And sorry for not mentioning it in the corresponding pull request :(

**Comment by huitema** [comment]
The code is set in `picoquic_prepare_packet_ready`, in  `sender.c`, lines 2095-2104:
```
if ((cnx->cnx_state == picoquic_state_server_false_start &&
        cnx->crypto_context[3].aead_decrypt != NULL) ||
        (cnx->cnx_state == picoquic_state_client_ready_start &&
            cnx->one_rtt_data_acknowledged)) {
        /* Transition to server ready state.
         * The handshake is complete, all the handshake packets are implicitly acknowledged */
        cnx->cnx_state = picoquic_state_ready;
        picoquic_implicit_handshake_ack(cnx, picoquic_packet_context_initial, current_time);
        picoquic_implicit_handshake_ack(cnx, picoquic_packet_context_handshake, current_time);
    }
```
The flag `one_rtt_data_acknowledged` is set in `picoquic_process_ack_range`, in `frame.c`, lines 1877. 

As far as I can tell, this works as expected. But there is a potential issue: it relies on the client sending 1-RTT data. If it did not, there would not be an ack. Do you have a scenario in which the client does not send any 1-RTT data?

**Comment by bkchr** [comment]
Hmm, in the current implementation it is not possible to send `1-RTT` data from rust. I should change this.
The rust library is written asynchronous and thus, I'm just resolving a new `Connection` as it is ready, to wait for stuff like failing certificates. 

But shouldn't the state switch to ready in `picoquic_process_ack_range` as he receives the corresponding data? Why should I need to start sending?

**Comment by huitema** [comment]
I will need to remember why I programmed the interaction between add and sending the way it is.

There may be something else that could be useful. How do you notice stste changes? Would a callback help?

**Comment by bkchr** [comment]
I think a callback would be really nice, currently I just check if the state is `ready` and if I have something that needs to be done on `ready`.

**Comment by huitema** [comment]
OK, let's keep this issue open with a request for callback. I am very busy developing the H3 responder required for interop tests, will schedule that work next.

**Comment by bkchr** [comment]
Yeah no rush :) But thank you as always! :)

**Comment by huitema** [comment]
Solved in PR #498
